### PR TITLE
docs: Add and format descriptions of the public API's class, methods, and parameters (EDU-2452)

### PR DESCRIPTION
**Pull request by jamesneimanconsulting**
## Summary
docs: Add and format descriptions of the public API's class, methods, and parameters (EDU-2452)
Provide additional detail to the descriptions and consistency with readme.io.

### CODE REVIEW

**Comment by jamesneimanconsulting** [comment]
@mikeng13 , @thomaszurkan-optimizely , @jonslaught I've submitted changes discussed in an additional commit.

File: packages/optimizely-sdk/lib/optimizely/index.js [hunk 1]
```diff
@@ -106,12 +112,18 @@ function Optimizely(config) {
...
- * Buckets visitor and sends impression event to Optimizely.
- * @param  {string}      experimentKey
- * @param  {string}      userId
- * @param  {Object}      attributes
- * @return {string|null} variation key
+/** 
+ * Activates an A/B test for a user, deciding whether they qualify for the experiment, bucketing them into a variation 
+ * if they do, and sending an impression event to Optimizely.
+ *
+ * For more information see [Activate](https://dash.readme.io/project/staging-optimizely-full-stack/v2/docs/instantiate).
```
Comment (mikeng13) [inline packages/optimizely-sdk/lib/optimizely/index.js]: This refers to the staging docs. There are other instances of this throughout the PR.

File: packages/optimizely-sdk/lib/optimizely/index.js [hunk 2]
```diff
@@ -106,12 +112,18 @@ function Optimizely(config) {
...
+ *
+ * @param  {string}      experimentKey  The key of the experiment for which to activate the variation.
+ * @param  {string}      userId         The ID of the user for whom to activate the variation.
+ * @param  {Object}      attributes     A map of up to 100 custom key-value string pairs defining attributes for the user. 
+ *
+ * @return {string|null} variation key  The key of the variation where the visitor will be bucketed., or `null` if the
```
Comment (mikeng13) [inline packages/optimizely-sdk/lib/optimizely/index.js]: nit: remove period after `bucketed`

File: packages/optimizely-sdk/lib/optimizely/index.js [hunk 3]
```diff
@@ -206,11 +220,16 @@ Optimizely.prototype._sendImpressionEvent = function(experimentKey, variationKey
...
- * @param  {Object} eventTags Values associated with the event.
+ * Tracks a conversion event for a user whose attributes meets the audience conditions for an experiment. 
+ * When the user does not meet those conditions, events are not tracked.
+ *
+ * For more information see [Track](https://dash.readme.io/project/staging-optimizely-full-stack/v2/docs/track).
+ *
+ * @param  {string} eventKey   The key of the event to be tracked. This key must match the event key provided when the event 
+ *                             was created in the Optimizely app.
+ * @param  {string} userId     The ID of the user associated with the event being tracked.
+ * @param  {string} attributes A map of up to 100 custom key-value string pairs specifying attributes for the user.
```
Comment (mikeng13) [inline packages/optimizely-sdk/lib/optimizely/index.js]: Where did the `100` limit come from? I am not sure we document this externally

File: packages/optimizely-sdk/lib/optimizely/index.js [hunk 4]
```diff
@@ -326,17 +359,19 @@ Optimizely.prototype.getVariation = function(experimentKey, userId, attributes)
...
+* The purpose of this method is to force a user into a specific variation or personalized experience for a given experiment.
+* The forced variation value does not persist across application launches.
+*
+* For more information see: https://docs.developers.optimizely.com/full-stack/docs/set-forced-variation.
+*
+* @param {string} experimentKey     The key of the experiment for which to set the forced variation.
+* @param {string} userId            The ID of the user for whom to set the forced variation.
+* @param {string|null} variationKey The ID of the variation to force the user into. Set the value to `null` to clear the existing experiment-to-variation mapping.
+*
+* @return boolean 'True' if the user was successfully forced into a variation, 'false' if the `experimentKey` is not in the project file or the `variationKey` is not in the experiment.
```
Comment (mikeng13) [inline packages/optimizely-sdk/lib/optimizely/index.js]: boolean `true`. In JS it's lowercase

File: packages/optimizely-sdk/lib/optimizely/index.js [hunk 5]
```diff
@@ -408,13 +436,15 @@ Optimizely.prototype.__validateInputs = function(stringInputs, userAttributes, e
...
- * @param  {string} eventKey
- * @param  {string} userId
- * @param  {Object} attributes
- * @return {Object} Map of experiment ids that we want to track to variations ids in which the user has been bucketed
+ *
+ * @param  {string} eventKey    The key of the experiment for which to retrieve experiments.
+ * @param  {string} userId      The ID of the user for whom to retrieve experiments.
+ * @param  {Object} attributes  A map of custom key-value string pairs specifying attributes for the user.
+ *
+ * @return {Object}             A map of experiment keys for which to track to variations in which the user has been bucketed.
```
Comment (jonslaught) [inline packages/optimizely-sdk/lib/optimizely/index.js]: "for which to track to" sounds like a typo

File: packages/optimizely-sdk/lib/optimizely/index.js [hunk 6]
```diff
@@ -656,18 +708,19 @@ Optimizely.prototype.getFeatureVariableBoolean = function (featureKey, variableK
...
- * @param {Object} attributes   Optional user attributes
- * @return {number|null}        Number value of the variable, or null if the
- *                              feature key is invalid, the variable key is
- *                              invalid, or there is a mismatch with the type
- *                              of the variable
+ * Evaluates and returns the value for the given double variable associated with a given feature.
+ * This method takes into account the user `attributes` passed in, to determine if the user
+ * is part of the audience that qualifies for the experiment.
+ *
+ * For more information see [Get Feature Variable](https://dash.readme.io/project/staging-optimizely-full-stack/v2/docs/get-feature-variable).
```
Comment (jonslaught) [inline packages/optimizely-sdk/lib/optimizely/index.js]: Another staging link

**Comment by jonslaught** [comment]
Sounds good thanks @jamesneimanconsulting !

**Comment by jennjblack** [comment]
Hi Jon,

James doesn't have access to the production site, so I'll take care of the
staging links.

Thanks,
jjb


Jennifer J. Black
*Senior Technical Writer*

Optimizely, Inc. <https://www.optimizely.com/>

jennifer.black@optimizely.com

**Comment by jonslaught** [comment]
These links should point to the public docs that everyone has access to,
otherwise our users won’t be able to follow them

**Comment by jamesneimanconsulting** [comment]
@jonslaught, @jennjblack  I'm re-checking this. I thought I had caught all the links that pointed to staging and revised them to point to production. I'll fix this in the next commit.

File: packages/optimizely-sdk/lib/optimizely/index.js [hunk 7]
```diff
@@ -106,12 +112,18 @@ function Optimizely(config) {
...
- * Buckets visitor and sends impression event to Optimizely.
- * @param  {string}      experimentKey
- * @param  {string}      userId
- * @param  {Object}      attributes
- * @return {string|null} variation key
+/** 
+ * Activates an A/B test for a user, deciding whether they qualify for the experiment, bucketing them into a variation 
+ * if they do, and sending an impression event to Optimizely.
+ *
+ * For more information see https://docs.developers.optimizely.com/full-stack/docs/activate.
```
[the remark this answers is not in the document]
Comment (jennjblack) [reply packages/optimizely-sdk/lib/optimizely/index.js]: Though the dev doc structure is relatively stable, these types of links will break over time. We'll need to add a process to manually check them until we get an automated link checker running. Though even if the link breaks and the user gets a 404, the link URL provides information about the original topic name; resourceful users will go searching in the doc for the info (while also logging a doc bug!).

**Comment by jamesneimanconsulting** [comment]
@jonslaught @jonslaught I've updated this PR with a new commit for consistency and a cross-project effort to align wording, grammar,  etc.

File: packages/optimizely-sdk/lib/optimizely/index.js [hunk 8]
```diff
@@ -312,20 +315,21 @@ Optimizely.prototype.track = function(eventKey, userId, attributes, eventTags) {
  * This method performs the same logic as `activate`, in that it activates an A/B test for
```
Comment (jennjblack) [inline packages/optimizely-sdk/lib/optimizely/index.js]: The `getVariation` method takes the same arguments and returns the same values as `activate`, but without sending an impression network request. The behavior of the two methods is identical otherwise.

File: packages/optimizely-sdk/lib/optimizely/index.js [hunk 9]
```diff
@@ -106,12 +112,18 @@ function Optimizely(config) {
...
- * @param  {Object}      attributes
- * @return {string|null} variation key
+/** 
+ * Activates an A/B test for a user, deciding whether they qualify for the experiment, bucketing them into a variation 
+ * if they do, and sending an impression event to Optimizely.
+ *
+ * For more information see https://docs.developers.optimizely.com/full-stack/docs/activate.
+ *
+ * @param  {string}      experimentKey  The key of the experiment for which to activate the variation.
+ * @param  {string}      userId         The ID of the user for whom to activate the variation.
```
[the remark this answers is not in the document]
Comment (jonslaught) [reply packages/optimizely-sdk/lib/optimizely/index.js]: I don't think I'd use the word "qualify" in that sense. But this also seems like overkill. "The user ID" would be a perfectly clear comment here, given the context above on how the function works.

File: packages/optimizely-sdk/lib/optimizely/index.js [hunk 10]
```diff
@@ -106,12 +112,18 @@ function Optimizely(config) {
...
+ * Activates an A/B test for a user, deciding whether they qualify for the experiment, bucketing them into a variation 
+ * if they do, and sending an impression event to Optimizely.
+ *
+ * For more information, see https://docs.developers.optimizely.com/full-stack/docs/activate.
+ *
+ * @param  {string}      experimentKey  The key of the experiment for which to activate the variation.
+ * @param  {string}      userId         The ID of the user for whom to activate the variation.
+ * @param  {Object}      attributes     A map of custom key-value string pairs specifying attributes for the user. 
+ *
+ * @return {string|null} variation key  The variation where the visitor will be bucketed, or `null` if the
```
[the remark this answers is not in the document]
Comment (jonslaught) [reply packages/optimizely-sdk/lib/optimizely/index.js]: Or simply "The bucketed variation, or..."

**Comment by jamesneimanconsulting** [comment]
@mikeng13 , @jennjblack, @jonslaught  I've updated all 5 repos based on all technical feedback (Jon, Mike) and stylistic change recommendations (Jennifer). Kindly let me know next steps.

**Comment by jonslaught** [comment]
@jamesneimanconsulting I've left a few more comments in the ruby-sdk PR. Once those are incorporated across all 5 languages, you have my approval to merge in. Not sure if JJB had any remaining open issues.

**Comment by mikeng13** [comment]
@jamesneimanconsulting can you merge master into your repo and update this PR

**Comment by jamesneimanconsulting** [comment]
@mikeng13 I have merged master into repo. In the process of updating all PRs per @jonslaught comment on the ruby sdk.

File: packages/optimizely-sdk/lib/optimizely/index.js [hunk 11]
```diff
@@ -35,17 +35,23 @@ var MODULE_NAME = 'OPTIMIZELY';
...
+/** @public
+ * @class
+ * The Optimizely client class containing an API to programmatically interact with Optimizely.
+ * The constructor accepts a configuration object to configure Optimizely.
+ * For more information see https://docs.developers.optimizely.com/full-stack/docs/instantiate.
+ *
+ * @param {Object} config                     Contains information to configure Optimizely.
+ * @param {string} config.clientEngine        Specifies the type of JavaScript used by the client. The value for this field must 
+ *                                            be set to "enums.NODE_CLIENT_ENGINE" or "enums.JAVASCRIPT_CLIENT_ENGINE".
+ * @param {string} config.clientVersion       Specifies the version of the client's JavaScript engine.
```
[the remark this answers is not in the document]
Comment (mikeng13) [reply packages/optimizely-sdk/lib/optimizely/index.js]: Yeah you are right, they should just be optional as we don't expect users to set them.

File: packages/optimizely-sdk/lib/optimizely/index.js [hunk 12]
```diff
@@ -496,11 +532,19 @@ Optimizely.prototype.__filterEmptyValues = function (map) {
 };
 
 /**
- * Returns true if the feature is enabled for the given user.
- * @param {string} featureKey   Key of feature which will be checked
- * @param {string} userId       ID of user which will be checked
- * @param {Object} attributes   Optional user attributes
- * @return {boolean}            True if the feature is enabled for the user, false otherwise
+ * Determines whether a feature test or rollout is enabled for a given user, and sends
+ * an impression event if the user is bucketed into an experiment using the feature.
```
Comment (mikeng13) [inline packages/optimizely-sdk/lib/optimizely/index.js]: Please note in the doc that feature test takes priority over the rollout.

Further down... `bucketed into a feature test using the feature. No impression event is sent if the user is bucketed into a rollout.`

**Comment by jamesneimanconsulting** [comment]
@jonslaught I've updated the branch. For the "config.clientEngine" and "config.clientVersion" parameters, according to http://usejsdoc.org/tags-param.html, an optional parameter in JS in denoted by wrapping the parameter name "[" and "]", so I've wrapped "config.clientEngine" and "config.clientVersion" accordingly. As far as I can tell this change is only required for JS and not the other languages.

File: packages/optimizely-sdk/lib/core/decision_service/index.js [hunk 13]
```diff
@@ -465,14 +508,16 @@ DecisionService.prototype._getBucketingId = function(userId, attributes) {
...
+   * @param  {Object} options.userProfileService An instance of the User Profile Service for sticky bucketing.
+   * @param  {Object} options.logger        An instance of a logger used to log messages.
+   *
+   * @return {Object} An instance of the decision service.
    */
   createDecisionService: function(options) {
     return new DecisionService(options);
   },
-};
+};
```
Comment (mikeng13) [inline packages/optimizely-sdk/lib/core/decision_service/index.js]: Please add missing newline at end of file

File: packages/optimizely-sdk/lib/optimizely/index.js [hunk 14]
```diff
@@ -347,16 +383,19 @@ Optimizely.prototype.setForcedVariation = function(experimentKey, userId, variat
...
+ * For more information, @see (@link https://docs.developers.optimizely.com/full-stack/docs/get-forced-variation}.
+ *
+ * @param  {string} experimentKey  The key of the experiment for which to retrieve the forced variation.
+ * @param  {string} userId         The ID of the user in the forced variation.
+ *
+ * @return {string|null}           The variation the user was bucketed into, or `null` if `setForcedVariation` 
+ *                                 failed to force the user into the variation.
 */
 Optimizely.prototype.getForcedVariation = function(experimentKey, userId) {
-  if (!this.__validateInputs({ experiment_key: experimentKey, user_id: userId })) {
```
Comment (mikeng13) [inline packages/optimizely-sdk/lib/optimizely/index.js]: This is likely the result of a bad merge. Please bring this back.

File: packages/optimizely-sdk/lib/optimizely/index.js [hunk 15]
```diff
@@ -367,25 +406,16 @@ Optimizely.prototype.getForcedVariation = function(experimentKey, userId) {
...
+ * @param  {string}  stringInputs   A map of string keys and associated values.
+ * @param  {Object}  userAttributes Optional user attributes to validate.
+ * @param  {Object}  eventTags      Optional event tags to validate.
  *
+ * @return {boolean}                'true' if all of the inputs are valid, 'false' if any of the inputs are invalid.
  */
 Optimizely.prototype.__validateInputs = function(stringInputs, userAttributes, eventTags) {
   try {
-    // Null, undefined or non-string user Id is invalid.
-    if (stringInputs.hasOwnProperty('user_id')) {
```
Comment (mikeng13) [inline packages/optimizely-sdk/lib/optimizely/index.js]: This shouldn't be removed.

File: packages/optimizely-sdk/lib/optimizely/index.js [hunk 16]
```diff
@@ -554,10 +609,6 @@ Optimizely.prototype.getEnabledFeatures = function (userId, attributes) {
       return enabledFeatures;
     }
 
-    if (!this.__validateInputs({ user_id: userId })) {
```
Comment (mikeng13) [inline packages/optimizely-sdk/lib/optimizely/index.js]: Revert this change

**Comment by jamesneimanconsulting** [comment]
@mikeng13 I've resolved the issues you raised. Kindly merge this into master.

File: packages/optimizely-sdk/lib/optimizely/index.js [hunk 17]
```diff
@@ -326,17 +359,20 @@ Optimizely.prototype.getVariation = function(experimentKey, userId, attributes)
...
+* @param {string} experimentKey     The key of the experiment to set with the forced variation.
+* @param {string} userId            The ID of the user to force into the variation.
+* @param {string|null} variationKey The key of the forced variation. 
+*                                   Set the value to `null` to clear the existing experiment-to-variation mapping.
+*
+* @return {boolean}                 `true` if the user was successfully forced into a variation, 
+*                                   `false` if the `experimentKey` isn't in the project file or the `variationKey` isn't in the experiment.
 */
 Optimizely.prototype.setForcedVariation = function(experimentKey, userId, variationKey) {
-  if (!this.__validateInputs({ experiment_key: experimentKey, user_id: userId })) {
```
Comment (mikeng13) [inline packages/optimizely-sdk/lib/optimizely/index.js]: This should not be changed

**Comment by mikeng13** [comment]
Obsolete now
